Documentation for graph classes (#190)
UCL/causalprog: docs/users/graph.md

> # Creating a using graphs in causalprog

Reviewer: ```suggestion
# Creating and using graphs in causalprog
```

automatically added to the graph when the node is added. For example, in the
following snippet a data node representing a scalar and a continuous random
variable node are added to the graph.

```python
from causalprog.graph import DataNode, ContinuousRandomVariableNode

graph.add_node(DataNode(label="first_node"))
graph.add_node(ContinuousRandomVariableNode(label="X", parents=["first_node"]))
```

As `"first_node"` is a parent of the random variable node, the edge pointing
from `"first_node"` to `"X"` will automatically be added the graph in the
final line of this snippet.

## Nodes

This section of the documentation details the different types of graph node
available in causalprog.

### `ConstantNode`

A `ContantNode` is a node that represents a known constant value. These
nodes have two required keyword arguments that must be passed: a `label` for
the node and the `value` that the node represents:

```python
import jax.numpy as jnp
from causalprog.graph import ConstantNode

one = ConstantNode(label="one", value=1.0)
vector = ConstantNode(label="v", value=jnp.array([1.0, 1.0, 2.0]))
```

### `DataNode`

A `DataNode` is a node that represents a constant value that is not known

> when the node is created. These nodes has one required keyword arguments

Reviewer: ```suggestion
when the node is created. These nodes have one required keyword argument
```

that must be passed: a `label` for the node. They can take the `shape` of
the data that the node represents as an additional keyword argument, with
the default shape being `()` for a scalar value.

```python
import jax.numpy as jnp
from causalprog.graph import DataNode

scalar = DataNode(label="my_scalar")
vector = DataNode(label="my_vector", shape=(5, ))
matrix = DataNode(label="my_matrix", shape=(4, 2))
```

### Random variable nodes

> Random variable node represent random variables (RVs) in a causal problem.

Reviewer: ```suggestion
Random variable nodes represent random variables (RVs) in a causal problem.
```

There are two types of random variable node in causalprog:
`ContinuousRandomVariableNode` and `DiscreteRandomVariableNode`. Both of
these must be passed the a `label` for the node as a required keyword
argument, and can take a number of additional keyword arguments: the `shape`
of the data that the RV outputs, a function to `compute` the value of the RV
from the values of its parents, and a list of `parents` of the RV node.

> Discrete RV nodes must be passed an addition required keyword argument: a

Reviewer: ```suggestion
Discrete RV nodes must be passed an additional required keyword argument: a
```

list of possible `values` that the RV can output.

The `compute` function should take a single input, and will be passes a
dictionary containing the values of the parents of the node, with node

> labels as keys. It should return the value of the RV.

Reviewer: ```suggestion
The `compute` function should take two inputs; 

- The first being a dictionary containing the values of the parents of the node, with the keys corresponding to the parent node labels.
- The second being a dictionary containing the current values of the parameters that parametrize the `compute` function.

Typically, the `compute` function will be some kind of predictive model or neural network that predicts the value of the RV from the parent node values (first input) and the parameters - in this case the weights and biases - of the neural network (second input).
The [`FunctionalMLP` class](FIXME-link-to-Sams-docs) demonstrates one such set of `compute` functions.

`compute` should return the value of the RV, as computed / predicted using the inputs.
```

`compute` should be an `MLPAlias` right? So it should take 2 input arguments. Can include a X-ref here to Sam's docs on MLPs when ready so that you don't have to provide a concrete example here, and can even X-ref to the example in #187 where we use deterministic functions.

Author: Opened #197 to remind us to add the link suggested here

```python
from causalprog.graph import ContinuousRandomVariableNode, DiscreteRandomVariableNode

node1 = DiscreteRandomVariableNode(values=[1.0, 1.5, 2.0], label="X")
node2 = ContinuousRandomVariableNode(label="Y")
node3 = ContinuousRandomVariableNode(
    label="2Y",

>     compute=lambda values: values["Y"] * 2,

Reviewer: ```suggestion
    compute=lambda values, parameters: values["Y"] * parameters["scale"],
```

To be consistent with the notation above.

    parents=["Y"],
)
```

### `DistributionNode`

Distrubution nodes represent the values of random variables (RVs) that can
be sampled from. These nodes formed part of an earlier experimental version
of the library and are not used in the current demonstration applications.

> ## Ricardo's graph

Reviewer: ```suggestion
## Continuous Treatment models
```

Just to move away from our "ricardo" terminology everywhere 😅

> Many of the examples in causalprog use an example graph for a problem

Reviewer: ```suggestion
Many of the examples in causalprog use an example graph, representing a continuous treatment model, for a problem
```

proposed by Ricardo Silva:

![Illustration of the continuous treatment model that we discuss.](../diagrams/continuous-treatment-model.svg)

causalprog provides a helper function to quickly generate this graph. This
function must be passed five required keyword arguments that tell the graph
how to compute the nodes `"u_x"`, `"u_y"`, `"phi_x"`, `"x"` and `"y"` from

> their parents.

Reviewer: May need to change this subject to how we resolve #156

```python
from causalprog.graph.ricardo import example_model

graph = example_model(
    compute_u_x=lambda values: values["c"][0] + 1.0,
    compute_u_y=lambda values: values["c"][1] * 2,
    compute_phi_x=lambda values: values["l"],
    compute_x=lambda values: values["z"] + values["phi_x"] - values["u_x"],

>     compute_y=lambda values: values["x"] * values["u_y"],

Reviewer: As above, these need to take 2 inputs. Though using `lambda values, _:` in each would solve this if you wanted a fast out.

Better would be to have them all of the form `lambda values, theta_{node_name}:` to be consistent with our examples.

Author: I'm going to update this in the next PR, as a few tests will need a matching update and it might make a mess.

Author: Issue for this: #196

)
```

## Using a graph

> This section of the documentation demonstrated how graphs created using

Reviewer: ```suggestion
This section of the documentation demonstrates how graphs created using
```

thought of as the starting points of the graph, like the roots of a tree).
The DAGs represented by causalprog are not necessarily trees, so may contain
more than one root node. The roots nodes of a graph can be obtained using
the property `graph.root_nodes`, which returns a tuple of nodes.

The leaf nodes of a graph are the nodes with no children (these can be
thought of as the ending points of the graph, liks the leaves of a tree).
The leaf nodes of a graph can be obtained using the property
`graph.leaf_nodes`, which returns a tuple of nodes.

### Predecessors and successors

In a DAG, the predecessors of a node are the parents of that node, plus
those parents' parents, plus their parents, and so on. Similarly, the
successors of the node are the node's children, plus the children's
children, and so on. In `causalprog` dictionaries mapping each node onto
tuples of this predecessors and successors can be obtained using the
properties `graph.predecessors` and `graph.successors`.

### Ordered nodes

The property `graph.ordered_nodes` and the method
`graph.roots_down_to_outcome` can be used to obtain tuples of nodes ordered

> so that every node's parents appear before that node in the tuple. This

Reviewer: ```suggestion
so that; for every node $X$ in the graph, the parents of node $X$ appear before node $X$ in the returned tuple. This
```

Don't know if this is any clearer than what you already have though

that the do is applied to with a `ConstantNode` and removes any predecessors
of the node that no longer have any children in the updated graph.

This algorithm takes three positional arguments: the graph, the label of the
node the do is applied to, and the value imposed on that node. It may also
take a keyword argument: the label of the newly created graph.

```python
from causalprog.algorithms import do

new_graph = do(graph, "x", 3.0)
```

### `evaluate` and `evaluate_down_to`

The `evaluate` and `evaluate_down_to` algorithms evaluate the values of
nodes in the graph given the values of some nodes as provided by the user.
Each of these algorithms takes three arguments: the graph, the label of a
node, and the values of any given nodes. The `evaluate` algorithm will
return the evaluated value of the node whose label is passed in; the
`evaluate_down_to` algrithms returns that node's value plus the value of all
of its predecessors, stored as a dictionary with the node labels as keys.

```python
import jax.numpy as jnp
from causalprog.algorithms import evaluate

value = evaluate(
    graph,
    "x",
    {"l": jnp.array([2.0]), "z": jnp.array([1.0]), "c": 1.0},
)

> ```

Reviewer: Just realised that these methods will need to take a fourth argument, which is a dictionary whose keys are the node labels and whose values are the parameter values of that node. Again because of the 2-input requirement of the `compute` functions.

Author: Will update this as part of #196
